### DISCUSSION TITLE
Last-modified when using github actions to publish

### DISCUSSION THREAD

**Discussion by adityam** [quarto-dev/quarto-cli#5899] [Q&A] [answered]
### Description

I am using github actions to publish my website.

1. Website: https://adityam.github.io/stochastic-control/
2. Github action: https://github.com/adityam/stochastic-control/blob/quarto/.github/workflows/quarto-render.yml

In `_quarto.yml`, I have set 

```
format:
  html:
    published-title: Updated
    date: last-modified
```
but all my pages have today's date. 

I suspect that the last modified date is coming from the date the date when github CI clones the repo (and hence today's date) rather than when the file was checked into the repo (which is the true last-modified date). 

Is it possible to somehow use the date when the file was last committed to git, for example, similar to [what is available for hugo](https://gohugo.io/variables/git/)

**Comment by mcanouil** [comment]
`last-modified` is for the rendered document not the Quarto document.
Since you don't use `freeze`, everything is rendered every time, thus every render is the last modified time, because indeed, the HTML files are modified.

If you don't want this, use `freeze`, see <https://quarto.org/docs/projects/code-execution.html#freeze>.

**Reply by cderv** [reply]
> last-modified is for the rendered document not the Quarto document.

@mcanouil `last-modified` is the date of the input file.  When 
```
date: last-modified
```
we are looking for the last modified date of the input file using `Deno.statSync()`. So this is supposed to be the date of the Quarto document. 

`freeze` should not have an impact I believe, as it should just be for computation. I think the date will still be resolved for a `date: today` even when freezing. 

If this is not working like this, then I think this is not what we are expecting. I wanted to add precision on this to not be misleading on how `last-modified` is working. 

Regarding the specific github action rendering here, I believe @adityam found the solution as https://github.com/quarto-dev/quarto-cli/discussions/5899#discussioncomment-6214370 shows 

the issue is that with Github Action, the project is checkout from github at each rendered, and unfortunately Git does not store the last modified date. This means the last modified date of the source is the same as the date of the rendering. 
This means `last-modified` is not working as expected in such render environment. 

Not that useful then... it will only works on project where source file are local and do not get their last modified date tweaked by Git. 

It requires a bit of processing using script as done in current workaround (there are helper tool for that too https://github.com/BenLubar/git-last-modified).

**Reply by mcanouil** [reply]
@cderv it makes sense, but the documentation is not that clear.

> The last modified date and time of the file containing the date.

See <https://quarto.org/docs/reference/dates.html>.

**Reply by cderv** [reply]
Yep sure no problem to rephrase. What would you add ?

**Reply by mcanouil** [reply]
"Input" since file can also refer to the rendered/output file. 😉

**Comment by adityam** [comment, marked as answer]
Thanks.

I could not get `freeze:auto` to work with github actions. I do not want to freeze the rendering ... freezing rendering could lead to stale links in cross refs (if the frozen page was referring to theorems/equations in other chapters). 

After a bit of googling, I found this script:

https://stackoverflow.com/questions/21735435/git-clone-changes-file-modification-time/55609950#55609950

Adding that as a github action does exactly what I want: it changes the mtime of all files to be the same as the last time that file was added to git.

**Reply by cderv** [reply]
@adityam I think you found the problem - I described it above https://github.com/quarto-dev/quarto-cli/discussions/5899#discussioncomment-6219611

I think we will try to improve this somehow in the future, especially for git project where we can do what you are doing. It could just slow things done so we may need to have a caching somehow... 

Note: Some CLI tools exists to help with this https://github.com/BenLubar/git-last-modified

**Reply by cderv** [reply]
Issue opened to remember and see what we can do: https://github.com/quarto-dev/quarto-cli/issues/5967

**Reply by adityam** [reply]
Thanks!

**Reply by lnnrtwttkhn** [reply]
hey @adityam, I have the same problem. could you point to an example how you implemented this in a github actions workflow? thanks!

**Reply by adityam** [reply]
See here: https://github.com/adityam/stochastic-control/blob/quarto/.github/workflows/quarto-render.yml

**Reply by mcanouil** [reply]
@adityam thank you for the link. Could you not use "here" to hide link, at least use plain URL or meaningful text. (This is an accessibility request.)

**Reply by adityam** [reply]
OK. I edited my reply with the whole link.
